content/courses/Ada_For_The_C_Embedded_Developer/chapters/06_Translation.rst

```diff
@@ -725,9 +725,467 @@ above could also be literally translated to:
+       begin
+          --  Assigning data from bit-field parameter B to local bit-field B_R.
+          --  R is automatically updated.
+          B_R := B;
+
+          return R;
+       end To_Rec;
+
+       procedure Display (R : Rec) is
+       begin
+          Put ("("  & Integer'Image (R.V) & ", "
+               & (R.S) & ")");
+       end;
+
+    end My_Recs;
+
+    with Ada.Text_IO; use Ada.Text_IO;
+    with Serializer;  use Serializer;
+    with My_Recs;     use My_Recs;
+
+    procedure Main is
+       R1 : Rec := (5, "abc");
+       R2 : Rec := (0, "zzz");
+
+       B1 : Bit_Field (0 .. R1'Size - 1) with Address => R1'Address;
+    begin
+       Put ("R2 = ");
+       Display(R2);
+       New_Line;
+
+       --  Getting Rec type using data from B1, which is a bit-field
+       --  representation of R1.
+       To_Rec (B1, R2);
+
+       --  We could use the function version of To_Rec:
+       --  R2 := To_Rec (B1);
+
+       Put_Line ("New bitstream received!");
+       Put ("R2 = ");
+       Display(R2);
+       New_Line;
+    end Main;
+
+In the procedure version of :ada:`To_Rec`, we create a bit-field representation
+of the output parameter (:ada:`B_R`) and simply copy the data from the input
+bit-field to :ada:`B_R`, so that the output parameter is updated. In the
+function version of :ada:`To_Rec`, however, we need to use a local object and a
+corresponding bit-field representation and return this object after updating
+the bit-field.
+
+In C, we can interpret the input pointer as an array of bytes, and copy the
+individual bytes. For example:
+
+[C]
+
+.. code:: c manual_chop run_button project=Courses.Ada_For_C_Embedded_Dev.Translation.Bitfield_Deserialization_C
+
+    !my_recs.h
+    typedef struct {
+        int v;
+        char s[3];
+    } rec;
+
+    void to_r (void *bits, int len, rec *r);
+
+    void display_r (rec *r);
+
+    !my_recs.c
+    #include "my_recs.h"
+
+    #include <stdio.h>
+    #include <assert.h>
+
+    void to_r (void *bits, int len, rec *r)
+    {
+        int i, j;
+        char *c1 = (char *)bits;
+        char *c2 = (char *)r;
+
+        assert(len == sizeof(rec) * 8);
+
+        for (i = 0; i < len / (sizeof(char) * 8); i++)
+        {
+            c2[i] = c1[i];
+        }
+    }
+
+    void display_r (rec *r)
+    {
+        printf("{%d, %c%c%c}", r->v, r->s[0], r->s[1], r->s[2]);
+    }
+
+    !bitfield_serialization.c
+    #include <stdio.h>
+    #include "my_recs.h"
+
+    int main(int argc, const char * argv[])
+    {
+        rec r1 = {5, "abc"};
+        rec r2 = {0, "zzz"};
+
+        printf("r2 = ");
+        display_r (&r2);
+        printf("\n");
+
+        to_r(&r1, sizeof(r1) * 8, &r2);
+
+        printf("New bitstream received!\n");
+        printf("r2 = ");
+        display_r (&r2);
+        printf("\n");
+    }
+
+Here, :c:`to_r` casts both pointer parameters to pointers to :c:`char` to get
+a byte-aligned pointer. Then, it simply copies the data byte-by-byte.
```

Comment:
I would suggest to add a section Overlays v.s. Unchecked_Conversions, explaining that Overlays are more efficient and allow to have one change automatically reflected by the other object, while Unchecked_Conversions require a copy. Unchecked_Conversions are not fundamentally bad and can also be used (they're actually cleaner and safer), but they're not always at the right level of performance / semantics. Perhaps that should be the introduction of this section.

Reply:
Yes, that makes sense. I'll add this topic to a separate section and work on it in another PR. Also, I'll add a link between the sections.